### PR TITLE
Fix Kali Linux support

### PR DESCRIPTION
**Pull request by koifans**
## Description

Current code doesn't really support Kali Linux. The reason is that it treats Kali Linux as a Debian distribution but it's actually different with Debian slightly -- the systemctl command takes different parameters than Debian. So the sshd daemon won't start after installation -- on Azure, the deployment will fail.

This patch provides one way to fix it. 

Issue # 1417 

This patch contains:
1. (changed) setup.py
    separate the workflow of Kali Linux handler from Debian
2. (changed) bin/waagent2.0
    provide placeholder for further dependency check
3. (changed) azurelinuxagent/common/osutil/factory.py
    reference to class KaliOSUtil
4. (new file) azurelinuxagent/common/osutil/kali.py
    add class KaliOSUtil
5. (new file) config/kali/waagent.conf
    configuration file for Kali Linux
6. (new file) config/kali/apt-setup
    workable apt-setup script on latest Kali Linux
7. (new file) init/kali/waagent-apt.service
    service file for systemctl command to setup apt
8. (new file) init/kali/walinuxagent.service
    service file for systemctl command to setup waagent
---

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1410)

### CODE REVIEW

**Comment by narrieta** [comment]
End to end automation is reporting errors on this branch... I need to investigate, will post an update.

**Comment by koifans** [comment]
I suppose to push to the master branch of **_koifans_**/WALinuxAgent and then merge into **_develop_** branch of **_Azure_**/WALinuxAgent. Now it says: **_continuous-integration/travis-ci Expected — Waiting for status to be reported_**, and somehow targeted master branch. Why?
How long it may take for the status to be updated? It's been almost a day. Any hint? Thanks.

**Comment by koifans** [comment]
Well... could someone please help to review this PR? Thanks. :)

**Comment by pgombar** [comment]
End-to-end automation is passing.

File: bin/waagent2.0 [hunk 1]
```diff
@@ -1436,6 +1436,9 @@ class KaliDistro(debianDistro):
     def __init__(self):
         super(KaliDistro,self).__init__()
 
+    def checkDependencies(self):
```
Comment (pgombar) [inline bin/waagent2.0]: Why is this needed?

File: config/kali/apt-setup [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+touch /var/lib/waagent/provisioned_apt-setup
+
+PROTOCOL=$(cat /var/lib/waagent/Protocol)
+
+if [ "$PROTOCOL" = "WireProtocol" ]; then
+	sed -ie 's#http://deb.debian.org#http://debian-archive.trafficmanager.net#' /etc/apt/sources.list
+fi
+
+apt-get update
```
Comment (pgombar) [inline config/kali/apt-setup]: Who will call this script? We have a known issue when the agent issues `apt-get update` since it can cause a circular dependency of the agent waiting on itself to update its own package.

File: config/kali/waagent.conf [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+# EnableOverProvisioning=y
+
+# Allow fallback to HTTP if HTTPS is unavailable
+# Note: Allowing HTTP (vs. HTTPS) may cause security risks
+# OS.AllowHTTP=n
+
+# Add firewall rules to protect access to Azure host node services
+# Note:
+# - The default is false to protect the state of existing VMs
+OS.EnableFirewall=n
```
Comment (pgombar) [inline config/kali/waagent.conf]: Can you please add the following entries:

```
# Enforce control groups limits on the agent and extensions
CGroups.EnforceLimits=n

# CGroups which are excluded from limits, comma separated
CGroups.Excluded=customscript,runcommand
```

File: init/kali/waagent-apt.service [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+#After=waagent-provision.service
+Before=multi-user.target
+ConditionVirtualization=microsoft
+
+[Service]
+ExecStart=/usr/share/waagent/apt-setup
+StandardOutput=journal+console
+
+[Install]
+WantedBy=walinuxagent.service
```
Comment (pgombar) [inline init/kali/waagent-apt.service]: Will this ensure that the `apt-setup` script is called before `walinuxagent.service` is run?

File: init/kali/walinuxagent.service [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#
+# NOTE:
+#   This file hosted on WALinuxAgent repository only for reference purposes.
+#   Please refer to a recent image to find out the up-to-date systemd unit file.
+#
+
+[Unit]
+Description=Azure Linux Agent
+After=multi-user.target
```
Comment (pgombar) [inline init/kali/walinuxagent.service]: Shouldn't `network-online.target` also be a prerequisite for the agent?

File: setup.py [hunk 6]
```diff
@@ -185,6 +194,12 @@ def get_data_files(name, version, fullname):
         set_sysv_files(data_files)
     return data_files
 
+def debian_has_systemd():
```
Comment (pgombar) [inline setup.py]: Duplicate of function directly below?

**Comment by narrieta** [comment]
sorry for the delay, please re-open if still active
